tech/post/2017-09-25-Kubernetes-Scheduler.md

```diff
@@ -0,0 +1,145 @@
+- First it applies a set of "predicates" to filter out inappropriate nodes. For example, if the PodSpec specifies resource requests, then the Scheduler will filter out nodes that don't have at least that much resources available.
+- Second, it applies a set of "priority functions" that rank the nodes that weren't filtered out by the predicate check. For example, it tries to spread Pods across nodes and zones while at the same time favoring the least loaded nodes.
+- Finally, the node with the highest priority is chosen, if there are multiple such nodes, then one of them is chosen at random.
+
+![](../images/2017-09-25-Kubernetes-Scheduler/Algorithm.png)
+
+## 3. Scheduler Code Analysis
+
+Scheduler main implementations include ```plugin/cmd/kube-scheduler``` and ```plugin/pkg/scheduler```.
+- ```plugin/cmd/kube-scheduler``` starts the Scheduler process and reads the configurations from Command Line Interface and so on.
+- ```plugin/pkg/scheduler/algorithm``` includes two kinds of scheduling algorithms: ```predicates``` and ```priorities```.
+- ```plugin/pkg/scheduler/algorithmprovider``` is used to register and configure Scheduler algorithm providers.
+- ```plugin/pkg/scheduler/factory``` creates scheduling algorithm and scheduling parameters.
+- ```plugin/pkg/scheduler/scheduler.go``` is the main class of Scheduler.
+- ```plugin/pkg/scheduler/core/generic_scheduler.go``` invokes the detail scheduling algorithm.
+
+The entrance of Scheduler is ```plugin/cmd/kube-scheduler/scheduler.go```.
+- First Scheduler creates ```configFactory``` and then creates ```config```.
+- If it specifies policy-config-file, Scheduler will create ```config``` by policy-config-file(CreateFromConfig),
+- If not, Scheduler will create ```config``` by Scheduler algorithm providers(CreateFromProvider),
+- After Scheduler have got all the parameters, the scheduling work will be executed in ```plugin/pkg/scheduler/scheduler.go```.
+- Scheduler invokes ```scheduleOne``` function and ```getNextPod``` from pod queue.
+- Then Scheduler execute ```schedule``` function, and actually the schedule process will be executed in ```plugin/pkg/scheduler/core/generic_scheduler.go```.
+- If the operation is successful, the Scheduler will get a most fit node for this pod.
+- Then Scheduler will tell the cache to assume that a pod now is running on a given node, even though it hasn't been bound yet. This allows Scheduler to keep scheduling without waiting on binding to occur. 
+- At last Scheduler will bind the pod to its node asynchronously by invoking Kubernetes APIServer Bind interface.
+
+![](../images/2017-09-25-Kubernetes-Scheduler/SchedulerSequence.png)
+
+The Kubernetes Scheduler is able to find the most fit node for pods. The actual scheduling steps include: ```predicates``` and ```priorities```. The default algorithm provider is ```DefaultProvider``` which can be specified by setting the parameter ```--algorithm-provider string``` when starting the Scheduler. 

```

Comment:
S/the most fit/the fittest

Reply:
ditto. http://fanyi.baidu.com/#en/zh/most%20fit